File: 2021/4/solve.py
INPUT_FILE = '2021/4/input.txt'

def input(name = INPUT_FILE):
    with open(name) as f:
        numbers = list(map(int, f.readline().strip('\n').split(',')))

        boards = []
        while f.readline():
            board = {}
            for l in range(5):
                for c, n in enumerate(map(int, f.readline().strip('\n').split())):
                    board[n] = (l, c)
            boards.append(board)

        return numbers, boards

def winners(numbers, boards):
    mark = tuple(tuple([0]*5 for _ in range(2)) for _ in boards)
    #print(mark)
    done = []
    wins = []
    for n in numbers:
        done.append(n)
        for i, b in enumerate(boards):
            if i in wins:
                continue
            if n in b:
                l, c = b[n]
                y = mark[i][0][l] + 1
                x = mark[i][1][c] + 1
                mark[i][0][l] = y
                mark[i][1][c] = x
                if x == 5 or y == 5:
                    #print('BINGO')
                    wins.append(i)
                    yield b, list(done), n

def solve():
    numbers, boards = input()

    for i, w in enumerate(winners(numbers, boards)):
        board, marked, n = w
        if i == 0:
            # part 1
            score = sum(k for k, _ in board.items() if not k in marked) * n
            print(score)

    # part 2
    score = sum(k for k, _ in board.items() if not k in marked) * n
    print(score)

solve()

File: 2021/4/solve.cpp
// requires -std=c++2a

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <tuple>
#include <sstream>
#include <set>
using namespace std;

tuple<int, int> EMPTY = {-1,-1};
typedef vector<tuple<int,int>> board_t;

tuple<vector<int>, vector<board_t>> readinput()
{
    string filename = "input.txt";
    string line, snum;
    vector<int> numbers;
    vector<board_t> boards;
    tie(numbers, boards);

    {
        ifstream file(filename, ios::in);
        getline(file, line);
        stringstream ss(line);
        while(getline(ss, snum, ','))
            numbers.push_back(stoi(snum));

        while(getline(file, line))
        {
            board_t board(100, EMPTY);
            for(int l=0; l<5; l++)
            {
                for(int c=0; c<5; c++)
                {
                    int n;
                    file >> n;
                    board[n] = {l, c};
                }
            }
            boards.push_back(board);
        }
    }

    return tie(numbers, boards);
}

int score(vector<bool>& done, board_t& board, int n)
{
    int s = 0;
    for(int i=0;i<100;i++)
        if(!done[i] && board[i] != EMPTY)
            s += i;

    return s * n;
}

int main()
{
    auto [numbers, boards] = readinput();

    int bs = boards.size();
    vector<vector<vector<int>>> mark(bs, vector<vector<int>>(2, vector<int>(5, 0)));
    vector<bool> done(numbers.size(), false);
    vector<bool> wins(bs, false);
    int rank = 0;
    for(int n: numbers)
    {
        done[n] = true;
        for(int i=0; i<bs; i++)
        {
            if(wins[i])
                continue;
            board_t& b = boards[i];
            if(b[n] == EMPTY)
                continue;
            auto [l, c] = b[n];
            int x = ++mark[i][0][l];
            int y = ++mark[i][1][c];
            if(x == 5 || y == 5)
            {
                wins[i] = true;
                if(rank == 0 || rank == bs-1)
                    cout << score(done, b, n) << endl;
                rank++;
            }
        }
    }

    return 0;
}
